Added ability to limit the amount of log messages stored

Overflwn opened the pull request:
The current version does not allow limiting the amount of log messages stored.

With bigger projects that run for hours or even days this will inevitably lead to memory usage inflation if the program logs a lot of messages over time.

arsdever commented on src/qspdlog_model.hpp:
```diff
@@ -32,4 +34,5 @@ class QSpdLogModel : public QAbstractListModel
 
 private:
     std::vector<entry_t> _items;
+    int _maxEntries;
```
I'm thinking if this could be an `std::optional<std::size_t>` instead.

Overflwn replied:
Honestly I never tried using std::optional and it looks like it really is useful in this situation, as you can now clearly see if there is a maximum or not.

arsdever commented on src/qspdlog_model.cpp:
```diff
@@ -22,20 +22,32 @@ static constexpr std::array<const char*, 3> column_names = { "Level",
 {
+    if(_maxEntries > 0 && _items.size() > _maxEntries) {
+        beginRemoveRows(QModelIndex(), 0, 0);
+        _items.erase(_items.begin());
```
`erase`-ing from the beginning of the `std::vector` can lead to performance degradation.

Also, as you see I map the `index`'s `row` to the n-th item in the vector in the `data` function. So, removing from the beginning for the vector will lead to incorrect indexation.

Overflwn replied:
> Also, as you see I map the index's row to the n-th item in the vector in the data function. So, removing from the beginning for the vector will lead to incorrect indexation.

I'm not exactly sure what you mean by that. You index `_items` in the `data` method via `index.row()`. If I understand the [QAbstractListModel](https://doc.qt.io/qt-6/qabstractitemmodel.html#beginRemoveRows) documentation correctly, `beginRemoveRows()` should handle relocating the indices.

Sorry if I'm wrong, I fairly new to Qt.

Overflwn commented:
I added the requested setter, removed maxEntries from the constructor, etc. like you requested.

I switched to a std::deque, I knew that std::vector::erase wasn't a good idea and I think deque is the proper collection type for it now.

arsdever commented on src/qspdlog_model.hpp:
```diff
@@ -19,6 +20,9 @@ class QSpdLogModel : public QAbstractListModel
     void addEntry(entry_t entry);
     void clear();
+    
+    std::optional<std::size_t> getMaxEntries() const;
+    void setMaxEntries(std::optional<std::size_t> maxEntries);
```
`set` goes before `get`, please.

arsdever commented on src/qspdlog_model.hpp:
```diff
@@ -31,5 +35,8 @@ class QSpdLogModel : public QAbstractListModel
 
 private:
-    std::vector<entry_t> _items;
+    //std::vector<entry_t> _items;
+    std::deque<entry_t> _items;
+    //int _maxEntries;
+    std::optional<std::size_t> _maxEntries;
```
Commented-out code should be removed if not explained why its there.

arsdever commented on src/qspdlog_model.cpp:
```diff
@@ -114,3 +126,6 @@ QVariant QSpdLogModel::headerData(
 
     return QVariant();
 }
+void QSpdLogModel::setMaxEntries(std::optional<std::size_t> maxEntries) {
+    _maxEntries = maxEntries;
+}
```
Move before the getter and keep the one-line margin between the function bodies.

arsdever commented on src/qspdlog.cpp:
```diff
@@ -153,3 +153,11 @@ std::size_t QSpdLog::itemsCount() const
 {
     return static_cast<std::size_t>(_proxyModel->rowCount());
 }
+
+std::optional<std::size_t> QSpdLog::getMaxEntries() const
+{
+  return _sourceModel->getMaxEntries();
+}
+void QSpdLog::setMaxEntries(std::optional<std::size_t> maxEntries) {
+  _sourceModel->setMaxEntries(maxEntries);
+}
```
Setter before the getter with one empty line margin in between.

arsdever commented on include/qspdlog/qspdlog.hpp:
```diff
@@ -87,6 +87,14 @@ class QSpdLog : public QWidget
+     * @return int the maximum number of items in the widget
+     */
+    std::optional<std::size_t> getMaxEntries() const;
+    void setMaxEntries(std::optional<std::size_t> maxEntries);
```
Move before the getter and add the documentation, please.

arsdever commented on src/qspdlog_model.cpp:
```diff
@@ -29,13 +29,25 @@ QSpdLogModel::QSpdLogModel(QObject* parent)
 {
+    if(_maxEntries > 0 && _items.size() > _maxEntries) {
+        beginRemoveRows(QModelIndex(), 0, 0);
+        //_items.erase(_items.begin());
```
No commented-out code, please.

arsdever reviewed (commented):
Thanks for your contribution. Overall looks good to me. Some style-related issues only.

Let's also have some unit tests covering the introduced changes.

Overflwn commented:
Thanks for the tips, I have applied them and also added a rudimentary unit test. I couldn't get your code coverage target to compile due to unknown compiler flags. Let me know if there's something missing.

I also extended the setMaxEntries() method to also remove excess items from the widget incase the new maximum is below the current amount of items.

arsdever commented:
> I couldn't get your code coverage target to compile due to unknown compiler flags. Let me know if there's something missing.

I'm not sure what environment you are working in, I'd like to know, though, to fix the incorrect setup (if there is any). The repo should just build and run for anybody who wants to.

I approved your PR and run the pipeline. As it turns out, the coverage is not that bad. The only function that's not covered though is the `getMaxEntry` function in both the model and the widget.

Overflwn commented:
> > I couldn't get your code coverage target to compile due to unknown compiler flags. Let me know if there's something missing.
> 
> I'm not sure what environment you are working in, I'd like to know, though, to fix the incorrect setup (if there is any). The repo should just build and run for anybody who wants to.
> 
> I approved your PR and run the pipeline. As it turns out, the coverage is not that bad. The only function that's not covered though is the `getMaxEntry` function in both the model and the widget.

I'm running Ubuntu 22.10 with qt6-base and libspdlog-dev installed, running `cmake .. -DBUILD_TESTING=ON -DBUILD_COVERAGE=ON` runs fine, but running `make` afterwards returns
```
[  0%] Built target qspdlog_interface
[  5%] Automatic MOC and UIC for target qspdlog_lib
[  5%] Built target qspdlog_lib_autogen
[ 11%] Automatic RCC for qspdlog_resources.qrc
[ 16%] Building CXX object src/CMakeFiles/qspdlog_lib.dir/qspdlog_lib_autogen/mocs_compilation.cpp.o
c++: error: unrecognized command-line option ‘-fprofile-instr-generate’; did you mean ‘-fprofile-generate’?
c++: error: unrecognized command-line option ‘-fcoverage-mapping’
make[2]: *** [src/CMakeFiles/qspdlog_lib.dir/build.make:88: src/CMakeFiles/qspdlog_lib.dir/qspdlog_lib_autogen/mocs_compilation.cpp.o] Fehler 1
make[1]: *** [CMakeFiles/Makefile2:182: src/CMakeFiles/qspdlog_lib.dir/all] Fehler 2
make: *** [Makefile:101: all] Fehler 2
```

Overflwn commented:
I added additional comparisons for the getMaxEntries() method inside the existing test, test ran fine on my machine.

arsdever commented:
> > > I couldn't get your code coverage target to compile due to unknown compiler flags. Let me know if there's something missing.
> > 
> > 
> > I'm not sure what environment you are working in, I'd like to know, though, to fix the incorrect setup (if there is any). The repo should just build and run for anybody who wants to.
> > I approved your PR and run the pipeline. As it turns out, the coverage is not that bad. The only function that's not covered though is the `getMaxEntry` function in both the model and the widget.
> 
> I'm running Ubuntu 22.10 with qt6-base and libspdlog-dev installed, running `cmake .. -DBUILD_TESTING=ON -DBUILD_COVERAGE=ON` runs fine, but running `make` afterwards (gcc version 12.2.0 (Ubuntu 12.2.0-3ubuntu1) ) returns
> 
> ```
> [  0%] Built target qspdlog_interface
> [  5%] Automatic MOC and UIC for target qspdlog_lib
> [  5%] Built target qspdlog_lib_autogen
> [ 11%] Automatic RCC for qspdlog_resources.qrc
> [ 16%] Building CXX object src/CMakeFiles/qspdlog_lib.dir/qspdlog_lib_autogen/mocs_compilation.cpp.o
> c++: error: unrecognized command-line option ‘-fprofile-instr-generate’; did you mean ‘-fprofile-generate’?
> c++: error: unrecognized command-line option ‘-fcoverage-mapping’
> make[2]: *** [src/CMakeFiles/qspdlog_lib.dir/build.make:88: src/CMakeFiles/qspdlog_lib.dir/qspdlog_lib_autogen/mocs_compilation.cpp.o] Fehler 1
> make[1]: *** [CMakeFiles/Makefile2:182: src/CMakeFiles/qspdlog_lib.dir/all] Fehler 2
> make: *** [Makefile:101: all] Fehler 2
> ```

Please open a separate issue for this one.
